src/Microsoft.AspNetCore.Razor.Language/VirtualRazorProjectFileSystem.cs

```diff
@@ -0,0 +1,200 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+
+namespace Microsoft.AspNetCore.Razor.Language
+{
+    internal class VirtualRazorProjectFileSystem : RazorProjectFileSystem
+    {
+        private readonly DirectoryNode Root = new DirectoryNode("/", "/");
+
+        public override IEnumerable<RazorProjectItem> EnumerateItems(string basePath)
+        {
+            basePath = NormalizeAndEnsureValidPath(basePath);
+            var directory = Root.GetDirectory(basePath);
+            return directory?.EnumerateItems() ?? Enumerable.Empty<RazorProjectItem>();
+        }
+
+        public override RazorProjectItem GetItem(string path)
+        {
+            path = NormalizeAndEnsureValidPath(path);
+            return Root.GetItem(path) ?? new NotFoundProjectItem(string.Empty, path);
+        }
+
+        public void Add(RazorProjectItem projectItem)
```

Comment:
Just to note (in regards to other comments I've left): if you were to lift `DirectoryNode` into its own class as `VirtualFileSystemHierarchy` you could easily test each piece individually in true unit test fashion 😄.

Reply:
null check project item.

Reply:
I'd rather leave `DirectoryNode` an implementation detail for `VirtualRazorProjectFileSystem`. We can still test these regardless.